### PR TITLE
[Docs] Improved nested tasks pattern example

### LINKED ISSUE #43892: [docs][core] Improve nested tasks pattern example
### Description

Our [nested tasks example](https://docs.ray.io/en/latest/ray-core/patterns/nested-tasks.html) is pretty simplistic and doesn't cover common edge cases like:
- ending up with too many parent tasks launched, and not enough cores for children
- copying data from parent task to child task instead of passing through an ObjectRef inside of a list (e.g., #43624).

### Link

_No response_

### PR DESCRIPTION
**Pull request by RajeevDash007**
## Why are these changes needed?

To address the issues mentioned and enhance the nested tasks example, I have  made the following adjustments:

1. Implementing a dynamic strategy to decide when to create new tasks based on the available resources and workload.
2. Passing data between tasks efficiently using ObjectRefs instead of copying data.
3. Ensuring proper resource management to avoid running out of cores for child tasks.

This updated code introduces dynamic resource allocation using ray.dynamic_resources() to manage resources more efficiently. It also ensures that ObjectRefs are passed between tasks instead of copying data. These improvements should help mitigate issues related to resource management and data copying, providing a more robust and efficient nested parallelism solution.
## Related issue number
closes #43892

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by RajeevDash007** [comment]
@stephanie-wang kindly check the changes,
and let me know if any more changes needed

**Review by aslonnie — COMMENTED** [review]
@jjyao could you have some one from core team to review this?

File: doc/source/ray-core/doc_code/pattern_nested_tasks.py [hunk 1]
```diff
@@ -49,11 +57,18 @@ def quick_sort_distributed(collection):
     s = time.time()
     quick_sort(unsorted)
     print(f"Sequential execution: {(time.time() - s):.3f}")
+
     s = time.time()
-    ray.get(quick_sort_distributed.remote(unsorted))
+    # Adjust resources dynamically based on workload and available cores
+    with ray.dynamic_resources(cpu=0):
```
Comment (jjyao) [inline doc/source/ray-core/doc_code/pattern_nested_tasks.py:63]: what's ray.dynamic_resources()? I don't think Ray has such method

**Comment by anyscalesam** [comment]
@RajeevDash007 can you please address @jjyao 's question?

**Pull request closed without merging**